### ISSUE TITLE
Why discriminator copy and two output of the generator? 

### ISSUE THREAD

**Issue by abodh** [daggertye/GAN-Q-Learning#2]
I have a few questions regarding your implementation. 

1. Why did you use the discriminator copy network? I have read somewhere in the paper that said a second network usage for stable training but the paper states that the second network is the target network (which is G'). But why did you use two discriminators?

2. Why is the output of generator 2? Is the generator giving the output as (s, a) pairs?

3. Why are you expanding the reward while giving it as an input to the discriminator?

**Comment by daggertye** [comment]
1. The second discriminator just reuses the weights of the first. To reuse the weights, I had to create a second "identical" network. These weights are needed to calculate various gradients.

2.  At a given state, the generator generates the next state and reward.

3. The discriminator is fed a reward as we take the next state (from the buffer), use the gen to get the reward. We don't directly use the state function since we need to perform a gradient update on both the gen and dis weights separately.

**Comment by abodh** [comment]
> 1. The second discriminator just reuses the weights of the first. To reuse the weights, I had to create a second "identical" network. These weights are needed to calculate various gradients.
> 2. At a given state, the generator generates the next state and reward.
> 3. The discriminator is fed a reward as we take the next state (from the buffer), use the gen to get the reward. We don't directly use the state function since we need to perform a gradient update on both the gen and dis weights separately.

Thank you so much for your response. I might have a few more questions while implementing this. Hope you do not mind me constantly posting my queries.

**Comment by abodh** [comment]
#The gradient for the loss function
    grad_val_ph = tf.placeholder(tf.float32, shape=dis.input_reward.get_shape())
    grad_dis = dis_copy(dis, grad_val_ph)
    #The generator-discriminator for loss function
    gen_dis = dis_copy(dis, tf.reduce_max(gen.output, axis=1))
    #loss functions
    dis_loss = tf.reduce_mean(tf.squeeze(gen_dis.output)) - tf.reduce_mean(tf.squeeze(dis.output)) \
            + lambda_ * tf.reduce_mean(tf.square(tf.gradients(grad_dis.output, grad_val_ph)[0] - 1))

    gen_loss = tf.reduce_mean(-tf.squeeze(gen_dis.output))

    #optimization
    optim = optimizer(learning_rate=learning_rate)
    dis_min_op = optim.minimize(dis_loss, var_list=dis.trainable_variables)
    gen_min_op = optim.minimize(gen_loss, var_list=gen.trainable_variables)


Hello Aaron,

Can you show the relation of the above part of the code with the original formula for loss used in the paper? 

also, I have another question:

action_results = sess.run(gen.output, feed_dict={
                gen.input_state : np.array([last_obs]),
                gen.input_seed : np.array([gen_seed])
            })[0]
            optimal_action = np.argmax(action_results)

why are you using action_results[0]? As you said,  the generator generates the next state and reward, which one is the state and which one is the reward? since it is a cart-pole problem, shouldn't there be 4 values as the state? I know these questions might seem stupid but I am having a hard time understanding this.

Thanks,
Abodh
